test/validation/api/ipsec/ipsec.c

```diff
@@ -101,11 +115,85 @@ static int pktio_start(odp_pktio_t pktio, odp_bool_t in, odp_bool_t out)
 	return 1;
 }
 
-static odp_event_t sched_queue_deq(uint64_t wait_ns)
+static int sched_event_cache_add(odp_queue_t from, odp_event_t event)
+{
+	odp_spinlock_lock(&sched_ev_cache_lock);
+
+	if (sched_ev_cache_tail + 1 == MAX_EVENT_CACHE) {
+		odp_spinlock_unlock(&sched_ev_cache_lock);
+
+		return -ENOMEM;
+	}
+
+	sched_ev_cache[sched_ev_cache_tail].from = from;
+	sched_ev_cache[sched_ev_cache_tail].event = event;
+	sched_ev_cache_tail++;
+
+	odp_spinlock_unlock(&sched_ev_cache_lock);
+
+	return 0;
+}
+
+static odp_event_t sched_event_cache_get(odp_queue_t from)
+{
+	odp_event_t ev;
+	int i, j;
+
+	if (odp_queue_type(from) == ODP_QUEUE_TYPE_PLAIN)
+		return ODP_EVENT_INVALID;
+
+	odp_spinlock_lock(&sched_ev_cache_lock);
+	if (sched_ev_cache_tail == 0) {
+		odp_spinlock_unlock(&sched_ev_cache_lock);
+
+		return ODP_EVENT_INVALID;
+	}
+
+	/* Look for a matching entry */
+	for (i = 0; i < sched_ev_cache_tail; i++)
+		if (sched_ev_cache[i].from == from)
+			break;
+
+	/* Remove entry from cache */
+	if (i != sched_ev_cache_tail) {
+		ev = sched_ev_cache[i].event;
+
+		for (j = 1; i + j < sched_ev_cache_tail; j++) {
+			sched_ev_cache[i + j - 1].from = sched_ev_cache[i + j].from;
+			sched_ev_cache[i + j - 1].event = sched_ev_cache[i + j].event;
```

Comment:
Struct assignment would be better and more robust, i.e. sched_ev_cache[i + j - 1] = sched_ev_cache[i + j];

Reply:
Ack. Using struct assignment now.